File: qh3server/qh3serverplugin_ts_proj/src/features/user_get/api_ping.ts
import { plainToClass } from 'class-transformer';
import { rq_msg_ping } from '../../userserver/messages';
import server from '../../userserver/userserver';

class api_ping implements server.interface_api {
    private static __LOGTAG__: string = `api_ping`;
    public get_path(): string {
        return `/ping`;
    }
    public get_post_cb(): (user_server_interface: server.interface_userserver, api_instance: server.interface_api, path: string, buffer: string, len: number) => Promise<string | null | any> {
        return this.parse_user_get;
    }
    private async parse_user_get(user_server_interface: server.interface_userserver, api_instance: server.interface_api, path: string, buffer: string, len: number) : Promise<string | null> {
        const ping_rq = plainToClass(rq_msg_ping, JSON.parse(buffer));
        const response_json = {
            pong:'qh3pluginserver',
            msg:ping_rq.msg
        };
        return JSON.stringify(response_json);
    }
}

export default api_ping;

File: serverplugin_lua_proj/src/features/api_ping.lua
-- local serversdk = require("src.helpers.serversdk")
local cjson = require("cjson")
local api_interface = require("src.features.api_interface")
local ffi = require("ffi")

-- Create the api_ping class
local api_ping = {}
api_ping.__index = api_ping
api_ping.__LOGTAG__ = "api_ping"

-- Inherit from the interface_api
setmetatable(api_ping, { __index = api_interface.interface_api })

-- Constructor
function api_ping:new()
    local self = setmetatable({}, api_ping)  -- Create an instance of api_ping
    -- Check if this class implements the interface
    -- api_interface.check_interface_implementation(self, api_interface.interface_api)
    return self
end

-- Implementing the get_path method required by the interface
function api_ping:get_path()
    return "/ping"
end

-- Implementing the get_post_cb method required by the interface
function api_ping:get_post_cb()
    return self.parse_ping
end

-- -- Implementing the parse_ping method
-- function api_ping:parse_ping(native_server, cid, cid_len, user_server_interface, api_instance, path, buffer, len, headers, header_buffer_size)
--     -- Assuming 'buffer' is the body of the incoming request as a JSON string
--     local ping_rq = cjson.decode(buffer)  -- Parse the JSON string into a Lua table

--     -- Assuming the `ping_rq` contains a `msg` field
--     local response_json = {
--         pong = "qh3pluginserver",  -- Static pong response
--         msg = ping_rq.msg          -- Echo the message from the incoming request
--     }

--     -- Convert the Lua table to a JSON string and return it
--     return cjson.encode(response_json)
-- end

api_ping.parse_ping = ffi.cast("type_on_api_parse", function(
    native_server, userserver_registry_id, native_cid, cid_len, native_path, native_buffer, len, native_headers_buffer, native_headers_buffer_size)
    local ping_rq = cjson.decode(ffi.string(native_buffer, len))  -- Parse the JSON string into a Lua table

    -- Assuming the `ping_rq` contains a `msg` field
    local response_json = {
        pong = "qh3pluginserver",  -- Static pong response
        msg = ping_rq.msg          -- Echo the message from the incoming request
    }

    -- Convert the Lua table to a JSON string and return it
    return cjson.encode(response_json)
end)

-- Return the class
return api_ping
